### ISSUE TITLE
Type inference keyword for lambda parameters

### ISSUE THREAD

**Issue by Zambonifofex** [ceylon/ceylon-spec#1391]
Currently, the following is allowed:
```Ceylon
function print(String string(Integer integer), Integer times) => // ...

print(function(i) => i.string, 12);
```
I think that instead it should be:
```Ceylon
print(function(value i) => i.string, 12);
```
And also:
```Ceylon
void foo(Bar bar(Baz baz())) => noop();

foo(function(function baz()) => nothing)
```
This would allow us to annotate a parameter and simultaneously have its type inferred, which is impossible today as far as I know, besides being more regular...

**Comment by gavinking** [comment]
@Zambonifofex do you have any usecases for annotating a parameter of an anonymous function?

**Comment by Zambonifofex** [comment]
@gavinking well, you can annotate it `variable` to be able to set it... And I suppose for documentation purposes... I think documentation is the only reason to actually allowing people to create their own annotations, is it not?

**Comment by gavinking** [comment]
Well a `variable` parameter is already pretty weird and I've never needed that in an anonymous function.

As for ceylondoc annotations I don't think they belong on the parameters of an anonymous function at all because (1) they would be ignored by ceylondoc (there is no doc generated for an anon function) and (2) adding doc strings and annotations in the middle of an expression seems just a bit awful.

There are *lots* of other uses for annotations, but they're uses that only really apply to APIs, not to anonymous functions which are fundamentally internal implementation details of a body of something.

So I guess I just don't see this.

**Comment by Zambonifofex** [comment]
No, I mean documentate it for people reading *the source code*. For example: Someone might make a `ignored` annotation for arguments that are ignored, and use it, so people don't get confused wondering *"What about this parameter? Did they just forget to use it?"*... I wasn't thinking about the `doc` annotation - I would also very much prefer to use comments to documentate the behavior of a lambda *if needed*, rather than to awkwardly fit a string in there... Either way, my main point is that it's more regular, at least in my opinion, to use a keyword to represent type inference, similar to how it's done with values/functions.

You don't do:
```Ceylon
foo = bar;
```
Instead, you do:
```Ceylon
value foo = bar;
```

Similarly, you don't do:
```Ceylon
foo(function(bar) => baz));
```
Instead, you do:
```Ceylon
foo(function(value bar) => baz));
```

**Comment by gavinking** [comment]
> No, I mean documentate it for people reading the source code.

Well that's what comments are for, no? :)

**Comment by gavinking** [comment]
> Either way, my main point is that it's more regular,

Ah I see. But:

    foo(function(bar) => baz));

is a shortcut for:

    foo(function(bar){ Bar bar; return baz; });

It's already regular :-)

**Comment by Zambonifofex** [comment]
Sure, but in my opinion, it's better to be explicit about the fact that you are declaring a new thing... At least *allowing* people to use a keyword can't be a bad thing, can it? :-P As well as allowing them to use the keyword if it's declared in the function body:
```Ceylon
foo(function(bar) { value bar; return baz; });
```
I don't know... I just don't like seeing `bar` sitting there by itself, representing a whole declaration... xP

**Comment by luolong** [comment]
The way I see this, it adding this little extra bit of explicit annotation is certainly more regular but that regularity does not buy us anything that would warrant the cost of more verbosity.

Ceylon has so far managed to strike a pretty good balance between verbose keyword heavy declarative syntax and concise expressive code.

I would argue that adding more verbose syntax to anonymous functions would tip that balance to the point where using anonymous functions becomes almost as verbose as declaring those functions on the top level explicitly and using function references explicitly. This is undesirable in my mind, as this will encourage wrong style of programming -- the code that makes sense only in the context of an anonymous function call site, will be divorced of it's context.

I'd say, the anonymous function syntax in Ceylon is really just as light weight as it needs to be and no lighter than it has to be.
